### ISSUE TITLE
Fatal error: 'ellipsoids.h' file not found

### ISSUE THREAD

**Issue by vaithak** [GeomScale/volume_approximation#62]
**Describe the bug**
Problem in installing the R-interface.

**To Reproduce**
I followed the given steps:
-> Install package-dependencies: Rcpp, RcppEigen, BH, lpSolveAPI
-> change working directory to "volume_approximation/R-Proj" using setwd
-> Ran these commands
```
Rcpp::compileAttributes()  
library(devtools)  
devtools::build()  
devtools::install()  
library(volesti)  
```

**Desktop:**
 - macOS Catalina

**Comment by vissarion** [comment]
Which of the following commands yield that error? Could you share more information about the error returned?

**Comment by vaithak** [comment]
I get the error after running devtools::install(), error log:
```
copula.cpp:11:ed from 
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/RcppEigen.h:25:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/RcppEigenForward.h:40:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/unsupported/Eigen/SparseExtra:51:
/Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/unsupported/Eigen/../../Eigen/src/Core/util/ReenableStupidWarnings.h:10:30: warning: pragma diagnostic pop could not pop, no matching push [-Wunknown-pragmas]
    #pragma clang diagnostic pop
                             ^
copula.cpp:13:10: fatal error: 'ellipsoids.h' file not found
#include "ellipsoids.h"
         ^~~~~~~~~~~~~~
17 warnings and 1 error generated.
make: *** [copula.o] Error 1
ERROR: compilation failed for package ‘volesti’
* removing ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library/volesti’
Error in (function (command = NULL, args = character(), error_on_status = TRUE,  : 
  System command error
```

I have not yet tried with CRAN package.

**Comment by vissarion** [comment]
Are trying to build the `develop` branch ? Which compiler/version? (it seems you are using `clang`)

**Comment by vaithak** [comment]
Yes, I am compiling the develop branch. 
Compiler: Apple clang version 11.0.0 (clang-1100.0.33.8)

**Comment by TolisChal** [comment]
Please open with R studio the file `volesti.Rproj` in 'root/R-proj` file to be sure that the working directory has been set correctly and then run:
`Rcpp::compileAttributes()  
library(devtools)  
devtools::build()  
devtools::install()  
library(volesti)`

Please report possible errors.

**Comment by vaithak** [comment]
Same errors:
```
In file included from copula.cpp:11:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/RcppEigen.h:25:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/RcppEigenForward.h:40:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/unsupported/Eigen/SparseExtra:51:
/Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/unsupported/Eigen/../../Eigen/src/Core/util/ReenableStupidWarnings.h:10:30: warning: pragma diagnostic pop could not pop, no matching push [-Wunknown-pragmas]
    #pragma clang diagnostic pop
                             ^
copula.cpp:13:10: fatal error: 'ellipsoids.h' file not found
#include "ellipsoids.h"
         ^~~~~~~~~~~~~~
17 warnings and 1 error generated.
make: *** [copula.o] Error 1
ERROR: compilation failed for package ‘volesti’
* removing ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library/volesti’
Error in (function (command = NULL, args = character(), error_on_status = TRUE,  : 
  System command error
```

**Comment by vaithak** [comment]
Hi, I tried the second way mentioned in README file by using **Build Source** and **Install and Restart** from build menu.
This time I get different errors:
```
clang -Wall -g -O2  -fPIC -I/usr/local/include -I../../../../external/LPsolve_src/include -I/Library/Frameworks/R.framework/Resources/include -DYY_NEVER_INTERACTIVE -DPARSER_LP -DINVERSE_ACTIVE=INVERSE_LUSOL -DRoleIsExternalInvEngine -c colamd.c -o colamd.o
colamd.c:772:5: error: unknown type name 'Colamd_Row'
    Colamd_Row Row [],
    ^
colamd.c:773:5: error: unknown type name 'Colamd_Col'
    Colamd_Col Col [],
    ^
colamd.c:783:5: error: unknown type name 'Colamd_Row'
    Colamd_Row Row [],
    ^
colamd.c:784:5: error: unknown type name 'Colamd_Col'
    Colamd_Col Col [],
    ^
colamd.c:798:5: error: unknown type name 'Colamd_Row'
    Colamd_Row Row [],
    ^
colamd.c:799:5: error: unknown type name 'Colamd_Col'
    Colamd_Col Col [],
    ^
colamd.c:810:5: error: unknown type name 'Colamd_Col'
    Colamd_Col Col [],
    ^
colamd.c:822:5: error: unknown type name 'Colamd_Col'
    Colamd_Col Col [],
    ^
colamd.c:833:5: error: unknown type name 'Colamd_Row'
    Colamd_Row Row [],
    ^
colamd.c:834:5: error: unknown type name 'Colamd_Col'
    Colamd_Col Col [],
    ^
colamd.c:842:5: error: unknown type name 'Colamd_Row'
    Colamd_Row Row []
    ^
colamd.c:954:12: error: conflicting types for 'colamd_recommended'
PUBLIC int colamd_recommended   /* returns recommended value of Alen. */
           ^
/usr/local/include/colamd.h:132:8: note: previous declaration is here
size_t colamd_recommended       /* returns recommended value of Alen, */
       ^
colamd.c:963:13: warning: implicit declaration of function 'COLAMD_RECOMMENDED' is invalid in C99 [-Wimplicit-function-declaration]
    return (COLAMD_RECOMMENDED (nnz, n_row, n_col)) ; 
            ^
colamd.c:1381:5: error: use of undeclared identifier 'Colamd_Row'
    Colamd_Row *Row ;           /* pointer into A of Row [0..n_row] array */
    ^
colamd.c:1381:17: error: use of undeclared identifier 'Row'
    Colamd_Row *Row ;           /* pointer into A of Row [0..n_row] array */
                ^
colamd.c:1382:5: error: use of undeclared identifier 'Colamd_Col'; did you mean 'colamd_l'?
    Colamd_Col *Col ;           /* pointer into A of Col [0..n_col] array */
    ^~~~~~~~~~
    colamd_l
/usr/local/include/colamd.h:169:18: note: 'colamd_l' declared here
SuiteSparse_long colamd_l       /* returns (1) if successful, (0) otherwise*/
                 ^
colamd.c:1382:17: error: use of undeclared identifier 'Col'
    Colamd_Col *Col ;           /* pointer into A of Col [0..n_col] array */
                ^
colamd.c:1465:16: warning: implicit declaration of function 'COLAMD_C' is invalid in C99 [-Wimplicit-function-declaration]
    Col_size = COLAMD_C (n_col) ;
               ^
colamd.c:1466:16: warning: implicit declaration of function 'COLAMD_R' is invalid in C99 [-Wimplicit-function-declaration]
    Row_size = COLAMD_R (n_row) ;
               ^
colamd.c:1480:5: error: use of undeclared identifier 'Col'
    Col = (Colamd_Col *) &A [Alen] ;
    ^
colamd.c:1480:24: error: expected expression
    Col = (Colamd_Col *) &A [Alen] ;
                       ^
colamd.c:1480:12: error: use of undeclared identifier 'Colamd_Col'; did you mean 'colamd_l'?
    Col = (Colamd_Col *) &A [Alen] ;
           ^~~~~~~~~~
           colamd_l
/usr/local/include/colamd.h:169:18: note: 'colamd_l' declared here
SuiteSparse_long colamd_l       /* returns (1) if successful, (0) otherwise*/
                 ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
3 warnings and 20 errors generated.
make[1]: *** [colamd.o] Error 1
make: *** [Rproj_externals/lp_solve/liblp_solve.a] Error 2
ERROR: compilation failed for package ‘volesti’
* removing ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library/volesti’

Exited with status 1.
```

**Comment by vaithak** [comment]
Hi, I tried building the package again, this time with gcc but still same error. I think this is an error in devtools which was introduced in macOS Mojave or Catalina.
A point to note is that, if I run the following command from bash in src folder, then there is no error, but the same command when ran through devtools::build gives error.
```
g++ -std=c++11 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../../external/boost -I../../external/LPsolve_src/run_headers -I../../external/minimum_ellipsoid -I../../include -I'/Library/Frameworks/R.framework/Versions/4.0/Resources/library/Rcpp/include' -I'/Library/Frameworks/R.framework/Versions/4.0/Resources/library/RcppEigen/include' -I'/Library/Frameworks/R.framework/Versions/4.0/Resources/library/BH/include' -I/usr/local/include  -lm -ldl -Wno-ignored-attributes -DBOOST_NO_AUTO_PTR -DDISABLE_NLP_ORACLES -fPIC  -Wall -g -O2  -c copula.cpp -o copula.o
```

I tried changing the -I path in Makevars file from relative to absolute path, then there was no error.
```
PKG_CPPFLAGS= -I$(HOME)/volume_approximation/external/boost -I$(HOME)/volume_approximation/external/LPsolve_src/run_headers -I$(HOME)/volume_approximation/external/minimum_ellipsoid -I$(HOME)/volume_approximation/include
```  

I think this can solve the issue but is there a way of using the absolute path without hardcoding it in the Makevars like above.

**Comment by vissarion** [comment]
@vaithak does https://github.com/GeomScale/volume_approximation/pull/137 solves your issue?

**Comment by vaithak** [comment]
I debugged this issue properly today, #137 helped a little, but the main issue was with macOS. Firstly, I had to follow the steps here: https://cmichel.io/fixing-cpp-compilation-bugs-for-the-mac-os-catalina-upgrade/ to fix the issue with Catalina. Other than this, GCC was not using the `colamd.h` in `external/LPsolve_src/include` but instead using the system installed one in `/usr/local/include` which had some issues. After fixing this, it worked fine :)

Closing the issue now.
